SwiftCompilerSources/Sources/Optimizer/Utilities/OwnershipUtils.swift

```diff
@@ -0,0 +1,159 @@
+//===--- OwnershipUtils.swift - Utilities for ownership -------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2023 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import SIL
+
+// Visit the introducers of a forwarded lifetime (Value -> ForwardIntroducers).
+//
+// A forward introducer represents an initial OSSA lifetime which may be extended by forwarding instructions. The introducer is never itself the result of a ForwardingInstruction. The introducer may be a phi or move_value. Although these conceptually forward values, they also summarize lifetime attributes and therefore are not formally considered ForwardingInstructions. (TODO: phi lifetime flags are not yet implemented).
+//
+// The lifetime of a single owned value ends when it is forwarded, but certain lifetime properties are relevant for the entire forward-extended lifetime. For example, if an owned lifetime has a pointer-escaping use, then all values in the forward-extended lifetime are also considered pointer-escaping. Certain propeties, like lexical lifetimes, only exist on the forward introducer and apply to all forwarded values.
+//
+// This walker is used to query basic lifetime attributes on values, such as "escaping" or "lexical". It must be precise for correctness and is performance critical.
+//
+// TODO: Ideally, this would be defined alongside ForwardingInstruction, but basic data structures and utilities are not available in the SIL module.
```

Comment:
Any walker that is required for well-formed SIL is part of SIL. In this case, the behavior of the walker will change SIL semantics, so it's definitely part of SIL. If the walker, "stops early" and we fail to see that some value has a pointer-escape, then we lose information needed for correctness. We also cache some information about lifetimes as instruction flags, and that cache needs to be consistent with the walker.

Reply:
Note that I'm not too concerned about the final resting place for new lifetime utilities because I expect them to change a lot in the next few months.